stage1/init/common/units.go

```diff
@@ -284,134 +283,121 @@ func (uw *UnitWriter) Error() error {
 }
 
 func (uw *UnitWriter) AppUnit(ra *schema.RuntimeApp, binPath string, opts ...*unit.UnitOption) {
-	if uw.err != nil {
+	if len(ra.App.Exec) == 0 {
+		uw.err = fmt.Errorf(`image %q has an empty "exec" (try --exec=BINARY)`,
+			uw.p.AppNameToImageName(ra.Name))
 		return
 	}
-
-	flavor, systemdVersion, err := GetFlavor(uw.p)
+	pa, err := prepareApp(uw.p, ra)
 	if err != nil {
-		uw.err = errwrap.Wrap(errors.New("unable to determine stage1 flavor"), err)
+		uw.err = err
 		return
 	}
 
-	app := ra.App
-	appName := ra.Name
-	imgName := uw.p.AppNameToImageName(appName)
-	imageManifest := uw.p.Images[appName.String()]
+	appName := ra.Name.String()
+	imgName := uw.p.AppNameToImageName(ra.Name)
+	/* Write the generic unit options */
+	opts = append(opts, []*unit.UnitOption{
+		unit.NewUnitOption("Unit", "Description", fmt.Sprintf("Application=%v Image=%v", appName, imgName)),
+		unit.NewUnitOption("Unit", "DefaultDependencies", "false"),
+		unit.NewUnitOption("Unit", "Wants", fmt.Sprintf("reaper-%s.service", appName)),
+		unit.NewUnitOption("Service", "Restart", "no"),
 
-	podAbsRoot, err := filepath.Abs(uw.p.Root)
-	if err != nil {
-		uw.err = err
-		return
-	}
+		// This helps working around a race
+		// (https://github.com/systemd/systemd/issues/2913) that causes the
+		// systemd unit name not getting written to the journal if the unit is
+		// short-lived and runs as non-root.
+		unit.NewUnitOption("Service", "SyslogIdentifier", appName),
+	}...)
 
-	if len(app.Exec) == 0 {
-		uw.err = fmt.Errorf(`image %q has an empty "exec" (try --exec=BINARY)`, imgName)
-		return
+	if supportsNotify(uw.p, ra.Name.String()) {
+		opts = append(opts, unit.NewUnitOption("Service", "Type", "notify"))
 	}
 
-	env := app.Environment
+	// When an app fails, we shut down the pod
+	opts = append(opts, unit.NewUnitOption("Unit", "OnFailure", "halt.target"))
 
-	env.Set("AC_APP_NAME", appName.String())
-	if uw.p.MetadataServiceURL != "" {
-		env.Set("AC_METADATA_URL", uw.p.MetadataServiceURL)
-	}
+	// Some pre-start jobs take a long time, set the timeout to 0
+	opts = append(opts, unit.NewUnitOption("Service", "TimeoutStartSec", "0"))
 
-	envFilePath := EnvFilePath(uw.p.Root, appName)
+	opts = append(opts, unit.NewUnitOption("Unit", "Requires", "sysusers.service"))
+	opts = append(opts, unit.NewUnitOption("Unit", "After", "sysusers.service"))
 
-	if err := common.WriteEnvFile(env, &uw.p.UidRange, envFilePath); err != nil {
-		uw.err = errwrap.Wrap(errors.New("unable to write environment file for systemd"), err)
+	opts = uw.AppSystemdUnit(pa, binPath, opts)
+
+	if uw.err != nil {
```

Comment:
This was structured the "errWriter" way, see https://blog.golang.org/errors-are-values, applied to the unitWriter.

The reasoning is that the caller can call unitWriter methods sequentially and check the error only once at the end of the sequence to avoid long `if err != nil ...` repetitions to the unit writer.

If we don't put the `if uw.err != nil` check at the top of the function, we break this semantic. I should and will add a reference to the godoc to the above mentioned article.